allennlp/modules/seq2seq_decoders/decoder_cell.py

```diff
@@ -0,0 +1,71 @@
+import torch
+from typing import Tuple
+
+from allennlp.common import Registrable
+
+
+class DecoderCell(torch.nn.Module, Registrable):
+    # pylint: disable=abstract-method
+    """
+    A ``DecoderCell`` is a ``Module`` that takes as input a result of ``Seq2SeqEncoder`` and returns a
+    new sequence of vectors.
+
+    Decoder should implement function to perform a single decoding step.
+
+    Parameters
+    ----------
+    decoding_dim : ``int``, required
+        Defines dimensionality of output vectors.
+
+    target_embedding_dim : ``int``, required
+        Defines dimensionality of target embeddings. Since this model takes it's output on a previous step
+        as input of following step, this is also an input dimensionality.
+    """
+
+    def __init__(self, decoding_dim, target_embedding_dim):
+        super(DecoderCell, self).__init__()
+        self.target_embedding_dim = target_embedding_dim
+        self._decoding_dim = decoding_dim
+
+    def get_output_dim(self) -> int:
+        """
+        Returns the dimension of each vector in the sequence output by this ``DecoderCell``.
+        This is `not` the shape of the returned tensor, but the last element of that shape.
+        """
+        return self._decoding_dim
+
+    def forward(self,
+                previous_steps_predictions: torch.Tensor,
+                encoder_outputs: torch.Tensor,
+                source_mask: torch.Tensor,
+                decoder_hidden: torch.Tensor,
+                decoder_context: torch.Tensor
+                ) -> Tuple[torch.Tensor, torch.Tensor]:
```

Comment:
I don't think this is the right return type - this assumes you have an LSTM.  You want this base class to be more general than that.  Unfortunately, this also affects the inputs to this function, too.  You could maybe get away with some generics here:

```python
# At the top of the file:
CellState = TypeVar('CellState')

# class definition
class DecoderCell(torch.nn.Module, Registrable, Generic[CellState]):

# at this line:
def forward(self,
            previous_steps_predictions: torch.Tensor,
            encoder_outputs: torch.Tensor,
            source_mask: torch.Tensor,
            previous_state: CellState) -> CellState:
```

Reply:
Current implementation of nn BeamSearch accepts states only as `Dict[str, torch.Tensor]`, so I we need decoder to have state of this type. It's steel pretty general, more general than a tuple